### PR TITLE
Encrypt and sign DCC (EXPOSUREAPP-10658)

### PR DESCRIPTION
**Pull request by chiljamgossow**
encryption of DCC and signing, needed before certificate can be submitted

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/core/security/DccTicketingSecurityTool.kt [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                iv = nonceBase64,
+                data = dccBarcodeData,
+                key = key
+            )
+        }
+    } catch (e: DccTicketingException) {
+        throw e
+    } catch (e: Exception) {
+        // anything else
+        throw DccTicketingException(DccTicketingException.ErrorCode.RSA_ENC_NOT_SUPPORTED)
```
Comment (jurajkusnier) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/core/security/DccTicketingSecurityTool.kt:45]: `encryptWithCBC()` and `encryptWithGCM()` throw only `DccTicketingException` 
shouldn't `RSA_ENC_NOT_SUPPORTED` be thrown by `input.encryptKey(key)` at least it mentioned like this in the spec https://github.com/corona-warn-app/cwa-app-tech-spec/blob/0f1e3a323fd2d52917ffc71ea256d26278e4e16b/docs/spec/dcc-validation-service.md#encrypting-data-with-rsa-oaep-with-sha256

Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/core/security/DccTicketingSecurityTool.kt:45]: True, that exception needs to be moved, thanks

**Review by jurajkusnier — COMMENTED** [review]
I don't see unit tests for 
- **Encrypting Data with RSA OAEP with SHA256** using [sample data from spec](https://github.com/corona-warn-app/cwa-app-tech-spec/blob/0f1e3a323fd2d52917ffc71ea256d26278e4e16b/docs/spec/dcc-validation-service.md#sample-data-for-encrypting-data-with-rsa-oaep-with-sha256) 
- **Signing Data with SHA256 and an EC Private Key** using [sample data from spec](https://github.com/corona-warn-app/cwa-app-tech-spec/blob/0f1e3a323fd2d52917ffc71ea256d26278e4e16b/docs/spec/dcc-validation-service.md#sample-data-for-signing-data-with-sha256-and-an-ec-private-key)

could you add?

**Comment by chiljamgossow** [comment]
> I don't see unit tests for
> 
> * **Encrypting Data with RSA OAEP with SHA256** using [sample data from spec](https://github.com/corona-warn-app/cwa-app-tech-spec/blob/0f1e3a323fd2d52917ffc71ea256d26278e4e16b/docs/spec/dcc-validation-service.md#sample-data-for-encrypting-data-with-rsa-oaep-with-sha256)

RSA was already in the app, I just added a test https://github.com/corona-warn-app/cwa-app-android/pull/4428/files#diff-1fc7016261c9378b94d9be14a4cf7705fb03cc69695affa7a19cdb9ead47b595R29 to confirm encryption and decryption results in original data

> * **Signing Data with SHA256 and an EC Private Key** using [sample data from spec](https://github.com/corona-warn-app/cwa-app-tech-spec/blob/0f1e3a323fd2d52917ffc71ea256d26278e4e16b/docs/spec/dcc-validation-service.md#sample-data-for-signing-data-with-sha256-and-an-ec-private-key)
> 
> could you add?

this one is here https://github.com/corona-warn-app/cwa-app-android/pull/4428/files#diff-6082c178b4d6d91d4a011a942224eb33efe1a5c7dfd756c096152be20b23c21bR16

**Comment by jurajkusnier** [comment]
> > I don't see unit tests for
> > 
> > * **Encrypting Data with RSA OAEP with SHA256** using [sample data from spec](https://github.com/corona-warn-app/cwa-app-tech-spec/blob/0f1e3a323fd2d52917ffc71ea256d26278e4e16b/docs/spec/dcc-validation-service.md#sample-data-for-encrypting-data-with-rsa-oaep-with-sha256)
> 
> RSA was already in the app, I just added a test https://github.com/corona-warn-app/cwa-app-android/pull/4428/files#diff-1fc7016261c9378b94d9be14a4cf7705fb03cc69695affa7a19cdb9ead47b595R29 to confirm encryption and decryption results in original data, but I can add it
> 
> > * **Signing Data with SHA256 and an EC Private Key** using [sample data from spec](https://github.com/corona-warn-app/cwa-app-tech-spec/blob/0f1e3a323fd2d52917ffc71ea256d26278e4e16b/docs/spec/dcc-validation-service.md#sample-data-for-signing-data-with-sha256-and-an-ec-private-key)
> > 
> > could you add?
> 
> this one is here https://github.com/corona-warn-app/cwa-app-android/pull/4428/files#diff-6082c178b4d6d91d4a011a942224eb33efe1a5c7dfd756c096152be20b23c21bR16

Right, I see it now, ok then it looks good :)

**Review by jurajkusnier — DISMISSED** [review]
LGTM

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/core/common/DccTicketingException.kt [hunk 2]
```diff
@@ -15,15 +15,17 @@ class DccTicketingException(
...
-*/
+
+        AES_CBC_INVALID_IV("AES CBC encryption iv parameter invalid."),
+        AES_CBC_INVALID_KEY("AES CBC encryption key is invalid."),
+        AES_CBC_NOT_SUPPORTED("AES CBC encryption algorithm is not supported."),
+        AES_GCM_INVALID_IV("AES GCM encryption iv parameter invalid."),
+        AES_GCM_INVALID_KEY("AES GCM encryption key is invalid."),
+        AES_GCM_NOT_SUPPORTED("AES GCM encryption algorithm is not supported."),
+        RSA_ENC_INVALID_KEY("RSA encryption key is invalid."),
+        RSA_ENC_NOT_SUPPORTED("RSA encryption algorithm is not supported."),
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/core/common/DccTicketingException.kt:26]: where 
```
EC_SIGN_INVALID_KEY
EC_SIGN_NOT_SUPPORTED
``` 
error codes should be handled or reported ? they are listed in the specs , but not used in here

Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/core/common/DccTicketingException.kt:26]: I forgot them, thanks

**Review by mtwalli — DISMISSED** [review]

**Review by mtwalli — APPROVED** [review]

**Review by jurajkusnier — APPROVED** [review]

**Pull request merged**